src/runtime/runtime_state.cs

```diff
@@ -0,0 +1,236 @@
+                    try
+                    {
+                        res = PySys_SetObject("initial_objs", objs);
+                        PythonException.ThrowIfIsNotZero(res);
+                    }
+                    finally
+                    {
+                        XDecref(objs);
+                    }
+                }
+            }
+        }
+
+        public static void Restore()
+        {
+            var dummyGCAddr = PySys_GetObject("dummy_gc");
+            if (dummyGCAddr == IntPtr.Zero)
+            {
+                throw new Exception("Runtime state have not set");
+            }
+            var dummyGC = PyLong_AsVoidPtr(dummyGCAddr);
+            ResotreModules(dummyGC);
+            if (ShouldRestoreObjects)
+            {
+                RestoreObjects(dummyGC);
+            }
+        }
+
+        private static void ResotreModules(IntPtr dummyGC)
+        {
+            var intialModules = PySys_GetObject("initial_modules");
+            Debug.Assert(intialModules != null);
+            var modules = PyImport_GetModuleDict();
+            foreach (var name in GetModuleNames())
+            {
+                if (PySet_Contains(intialModules, name) == 1)
+                {
+                    continue;
+                }
+                var module = PyDict_GetItem(modules, name);
+
+                if (UseDummyGC && _PyObject_GC_IS_TRACKED(module))
+                {
+                    ExchangeGCChain(module, dummyGC);
+                }
+                if (PyDict_DelItem(modules, name) != 0)
+                {
+                    PyErr_Print();
+                }
+            }
+        }
+
+        private static void RestoreObjects(IntPtr dummyGC)
+        {
+            if (!UseDummyGC)
+            {
+                throw new Exception("To prevent crash by _PyObject_GC_UNTRACK in Python internal, UseDummyGC should be enabled when using ResotreObjects");
+            }
+            var intialObjs = PySys_GetObject("initial_objs");
+            Debug.Assert(intialObjs != null);
+            foreach (var obj in PyGCGetObjects())
+            {
+                var p = PyLong_FromVoidPtr(obj);
+                try
+                {
+                    if (PySet_Contains(intialObjs, p) == 1)
+                    {
+                        continue;
+                    }
+                }
+                finally
+                {
+                    XDecref(p);
+                }
+                Debug.Assert(_PyObject_GC_IS_TRACKED(obj), "A GC object must be tracked");
+                ExchangeGCChain(obj, dummyGC);
+            }
+        }
+
+        public static IEnumerable<IntPtr> PyGCGetObjects()
+        {
+            var gc = PyImport_ImportModule("gc");
+            var get_objects = PyObject_GetAttrString(gc, "get_objects");
+            var objs = PyObject_CallObject(get_objects, IntPtr.Zero);
+            var length = PyList_Size(objs);
```

Comment:
Please, check the return values of these calls.

Also, wrap code, operating on `objs` and `gc` in `try .. finally ...`

Reply:
Yes, checking return values is necessary although using these builtin functions will be failed by less probability.
But the try is not necessary, they're all C functions actually.